src/timely-util/src/builder_async.rs

```diff
@@ -178,6 +178,18 @@ impl<'handle, T: Timestamp, D: Container, P: Pull<BundleCore<T, D>>> Future
         let state: &'handle mut NextFutState<T, D, P> =
             self.state.take().expect("future polled after completion");
 
+        // Check for frontier notifications first, to urge operators to retire pending work
+        let mut shared_frontiers = state.shared_frontiers.borrow_mut();
+        let (ref frontier, ref mut pending) = shared_frontiers[state.index];
+        if *pending {
+            *pending = false;
+            return Poll::Ready(Some(Event::Progress(frontier.clone())));
+        } else if frontier.is_empty() {
+            // If the frontier is empty and is not pending it means that there is no more data left
+            return Poll::Ready(None);
+        };
+        drop(shared_frontiers);
+
```

Comment:
if an upstream operator is moving the frontier very quickly, can this starve fetching the data below?

Reply:
No, frontiers move as a result of ChangeBatches getting into a `MutableAntichain` and if there are a thousand small batches between each invocation when you observe it you will consume all of it and then immediately start consuming the data

Reply:
It can't, timely workers are single threaded so as long as the operator is consuming the handle you will see at most one frontier

Reply:
@petrosagg ah yes!